docs/reference/server-events.md

```diff
@@ -0,0 +1,982 @@
+      }
+    };
+  }
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/page/index.js
+  </template>
+</AposCodeBlock>
+
+### `beforeMove`
+
+Triggered just *before* a page is moved from one position in the page tree to another.
+
+#### Parameters
+
+- `req`: The active request
+- `page`: The page being moved
+- `target`: The page used as a target for positioning in the page tree
+- `position`: The position to place the page in relation to the target
+
+<AposCodeBlock>
+  ```javascript
+  handlers(self, options) {
+    return {
+      'beforeMove': {
+        async handlerName(req, page, target, position) { ... }
+      }
+    };
+  }
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/page/index.js
+  </template>
+</AposCodeBlock>
+
+### `afterMove`
+
+Triggered just *after* a page is moved from one position in the page tree to another.
+
+#### Parameters
+
+- `req`: The active request
+- `page`: The page being moved
+- `data`: An object containing the following properties:
+  - `originalSlug`: The page's slug prior to moving
+  - `originalPath`: The page's path prior to moving
+  - `changed`: An array of all pages that were changed by the move, including children of the moved page
+  - `target`: The page used as a target for positioning in the page tree
+  - `position`: The position to place the page in relation to the target
+
+<AposCodeBlock>
+  ```javascript
+  handlers(self, options) {
+    return {
+      'afterMove': {
+        async handlerName(req, page, data) { ... }
+      }
+    };
+  }
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/page/index.js
+  </template>
+</AposCodeBlock>
+
+### `serveQuery`
+
+Triggered just before the database query executes to find the best matching page for a request. This can be used to make final adjustments to the query.
+
+#### Parameters
+
+- `req`: The active request
+- `query`: The database query that will execute
+
+<AposCodeBlock>
+  ```javascript
+  handlers(self, options) {
+    return {
+      'serveQuery': {
+        async handlerName(req, query) { ... }
+      }
+    };
+  }
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/page/index.js
+  </template>
+</AposCodeBlock>
+
+### `serve`
+
+Triggered just before a requested page is served.
+
+#### Parameters
+
+- `req`: The active request
+
+<AposCodeBlock>
+  ```javascript
+  handlers(self, options) {
+    return {
+      'serve': {
+        async handlerName(req) { ... }
+      }
+    };
+  }
+  ```
+  <template v-slot:caption>
+    modules/@apostrophecms/page/index.js
+  </template>
+</AposCodeBlock>
+
+### `notFound`
+
+Triggered when Apostrophe fails to find a matching page for a request, just before the 404 response. This is a final chance to do extra work to find a matching page.
+
+#### Parameters
+
+- `req`: The active request
+
+<AposCodeBlock>
+  ```javascript
+  handlers(self, options) {
+    return {
+      'notFound': {
+        async handlerName(req) { ... }
```

Comment:
Feeling the lack of an example of how to make the 404 not happen after all (see the soft-redirects module, I think).